Add access control feature

hyangtack opened the pull request:
Modifications:
- Do access control for every v1 APIs.
- Update metadata when do a CRUD operation for a project.
- Build metadata if no metadata exists.
- Administrator IDs can be specified in the configuration.
- Add web UI for metadata management.

To-Dos:
- Add more test cases and cleanup test cases.
- Documentation (site)

trustin commented on common/src/main/java/com/linecorp/centraldogma/internal/jsonpatch/RemoveIfExistsOperation.java:
```diff
@@ -0,0 +1,86 @@
+ * <p>This operation only takes one pointer ({@code path}) as an argument. It
+ * is an error condition if no JSON value exists at that pointer.</p>
+ */
+public final class RemoveIfExistsOperation extends JsonPatchOperation {
```
How about adding a test case?

hyangtack replied:
Added

trustin commented on common/src/main/java/com/linecorp/centraldogma/internal/jsonpatch/TestAbsenceOperation.java:
```diff
@@ -0,0 +1,62 @@
+import com.fasterxml.jackson.databind.SerializerProvider;
+import com.fasterxml.jackson.databind.jsontype.TypeSerializer;
+
+public final class TestAbsenceOperation extends JsonPatchOperation {
```
How about adding a test case?

hyangtack replied:
Added

trustin commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -3,4 +3,4 @@ distributionPath=wrapper/dists
 zipStorePath=wrapper/dists
 distributionUrl=https\://services.gradle.org/distributions/gradle-4.5.1-all.zip
-distributionSha256Sum=ede2333b509ab67e4b6f51adb1419870fb34bbead0c8357dd803532b531508eb
+distributionSha256Sum=ede2333b509ab67e4b6f51adb1419870fb34bbead0c8357dd803532b531508eb
```
Probably better reverting?

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/CentralDogmaBuilder.java:
```diff
@@ -272,6 +275,22 @@ public CentralDogmaBuilder securityConfig(Ini securityConfig) {
+     * Configures the list of administrators.
+     */
+    public CentralDogmaBuilder administrators(String... administrators) {
+        this.administrators = ImmutableSet.copyOf(administrators);
```
`requireNonNull(administrators, "administrators")`

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/CentralDogmaBuilder.java:
```diff
@@ -272,6 +275,22 @@ public CentralDogmaBuilder securityConfig(Ini securityConfig) {
+     * Configures the list of administrators.
+     */
+    public CentralDogmaBuilder administrators(Iterable<String> administrators) {
+        this.administrators = ImmutableSet.copyOf(administrators);
```
`requireNonNull(administrators, "administrators")`

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/CentralDogmaBuilder.java:
```diff
@@ -78,6 +80,7 @@
     private GracefulShutdownTimeout gracefulShutdownTimeout;
     private ReplicationConfig replicationConfig = ReplicationConfig.NONE;
     private Ini securityConfig;
+    private Set<String> administrators;
```
I'm curious if we should make this `final` and make `administrators()` always add, instead of setting:

```java
// foo and bar are both admins.
builder.admin('foo');
builder.admin('bar');
```

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/internal/api/MetadataApiService.java:
```diff
@@ -0,0 +1,393 @@
+        }
+    }
+
+    static class IdentifierWithPermissions {
```
final?

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/internal/api/MetadataApiService.java:
```diff
@@ -0,0 +1,393 @@
+
+    // TODO(hyangtack) Move these classes to the common module later when our java client accesses to
+    //                 the metadata.
+    static class IdentifierWithRole {
```
final?

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/internal/api/auth/AbstractPermissionCheckingDecorator.java:
```diff
@@ -0,0 +1,131 @@
+            if (user.isAdmin()) {
+                f = CompletableFuture.completedFuture(ADMIN_PERMISSION);
+            } else if (user instanceof UserWithToken) {
+                f = mds.findPermission(projectName, repoName, (UserWithToken) user);
```
How about moving `user instanceof UserWithToken` to `MetadataService.findPermissions()` so that a caller doesn't need to downcast explicitly?

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/internal/api/auth/ProjectOwnersOnly.java:
```diff
@@ -14,21 +14,22 @@
-        return user == User.ADMIN || projectRole == ProjectRole.OWNER;
+    protected boolean isAccessAllowed(ServiceRequestContext ctx, HttpRequest req,
+                                      User user, ProjectRole role) {
+        return user.roles() == User.LEVEL_ADMIN || role == ProjectRole.OWNER;
```
`User.roles()` returns `List<String>`, which means it will never be same with `User.LEVEL_ADMIN`?

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/internal/metadata/MetadataService.java:
```diff
@@ -0,0 +1,853 @@
+
+        final Change<JsonNode> change = Change.ofJsonPatch(
+                METADATA_JSON,
+                new ReplaceOperation(JsonPointer.compile("/members/" + member.id() + "/role"),
```
Perhaps we should put some validation on its `id` property when a `User` is created so we do not have to worry about potential security issues?

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/internal/metadata/MetadataService.java:
```diff
@@ -0,0 +1,853 @@
+     * Adds a collection of {@link Permission}s for the specified {@code member} to the specified
+     * {@code projectName} and {@code repoName}.
+     */
+    public CompletionStage<Revision> addSpecificUserPermission(Author author, String projectName,
```
`SpecificUser` -> `PerUser`?

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/internal/metadata/MetadataService.java:
```diff
@@ -0,0 +1,853 @@
+     * Updates a collection of {@link Permission}s for the specified {@code member} in the specified
+     * {@code projectName} and {@code repoName}.
+     */
+    public CompletionStage<Revision> updateUserSpecificPermission(Author author, String projectName,
```
`UserSpecific` -> `PerUser`?

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/internal/metadata/MetadataService.java:
```diff
@@ -0,0 +1,853 @@
+     * Adds a collection of {@link Permission}s for the {@link Token} of the specified {@code appId}
+     * to the specified {@code projectName} and {@code repoName}.
+     */
+    public CompletionStage<Revision> addSpecificTokenPermission(Author author, String projectName,
```
`SpecificToken` -> `PerToken`?

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/internal/metadata/MetadataService.java:
```diff
@@ -0,0 +1,853 @@
+    /**
+     * Adds {@link Permission}s to the specified {@code path}.
+     */
+    private CompletionStage<Revision> addSpecificPermission(Author author,
```
`SpecificPermission` -> `PermissionAtPointer`?

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/internal/metadata/MetadataService.java:
```diff
@@ -0,0 +1,853 @@
+     * {@link UserWithToken} from the specified {@code projectName} and {@code repoName}.
+     */
+    public CompletionStage<Collection<Permission>> findPermission(String projectName, String repoName,
+                                                                  UserWithToken userWithToken) {
```
This method could be merged into `findPermission(..., User)` with an `instanceof` check given `UserWithToken` is a subtype of `User`.

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/internal/metadata/MetadataService.java:
```diff
@@ -0,0 +1,853 @@
+                                      .thenApply(tokens -> {
+                                          final Token token = tokens.object().get(appId);
+                                          final JsonPointer removalPath =
+                                                  JsonPointer.compile("/appIds/" + appId + "/deactivation");
```
Same for `appId` - It would be great if we have a single place that validates all `appId` strings.

hyangtack replied:
Added validation for `appId` and `secret` to the constructor of `Token` class. I think it's enough to use `Util.validateFileName()` for these cases. Please let me know if you have any concern about that.

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/internal/metadata/Permission.java:
```diff
@@ -0,0 +1,31 @@
+ */
+public enum Permission {
+    /**
+     * Able to read files from a repository.
```
files -> a file

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/internal/metadata/RepositoryUtil.java:
```diff
@@ -0,0 +1,180 @@
+    }
+
+    @SuppressWarnings("unchecked")
+    static <T> T jacksonConverter(Entry<?> entry, Class<T> clazz) {
```
`convertWithJackson`?

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/internal/metadata/RolePermission.java:
```diff
@@ -0,0 +1,110 @@
+                          @JsonProperty("guest") Collection<Permission> guest) {
+        this.owner = requireNonNull(owner, "owner");
+        this.member = requireNonNull(member, "member");
+        this.guest = requireNonNull(guest, "guest");
```
Could be copied into an immutable `EnumSet`s?

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/internal/metadata/RolePermission.java:
```diff
@@ -0,0 +1,110 @@
+    /**
+     * Creates a {@link RolePermission} which allows accessing a repository from everyone.
+     */
+    public static RolePermission ofPublic() {
```
Please move the public static methods before non-static members.

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/internal/metadata/RolePermission.java:
```diff
@@ -0,0 +1,110 @@
+/**
+ * A default permission for a {@link Repository}.
+ */
+public class RolePermission {
```
How about `PerRolePermissions`? Can't think of a good name..

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/internal/metadata/Token.java:
```diff
@@ -0,0 +1,126 @@
+                 @JsonProperty("secret") String secret,
+                 @JsonProperty("creation") UserAndTimestamp creation,
+                 @JsonProperty("deactivation") @Nullable UserAndTimestamp deactivation) {
+        this.appId = requireNonNull(appId, "appId");
```
We could add some validation for `appId` and `secret`.

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/internal/metadata/Token.java:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    private Token(String appId, UserAndTimestamp creation, @Nullable UserAndTimestamp deactivation) {
+        this.appId = requireNonNull(appId, "appId");
```
and here as well.

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/internal/metadata/Tokens.java:
```diff
@@ -0,0 +1,140 @@
+                          .toString();
+    }
+
+    public Tokens duplicateWithoutTokenSecret() {
```
- Could be moved before `toString()`?
- How about just `withoutSecret()`?

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/internal/metadata/WithLocation.java:
```diff
@@ -0,0 +1,42 @@
+
+import static java.util.Objects.requireNonNull;
+
+public final class WithLocation<T> {
```
`HolderWithLocation`?

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/internal/metadata/WithRevision.java:
```diff
@@ -0,0 +1,44 @@
+
+import com.linecorp.centraldogma.common.Revision;
+
+final class WithRevision<T> {
```
`HolderWithRevision`?

trustin commented on site/src/sphinx/setup-configuration.rst:
```diff
@@ -167,6 +168,10 @@ Core properties
 
+- ``administrators`` (set of string)
+
+  - the set of administrator's ID. They are valid only if ``securityEnabled`` is ``true``.
```
the set of administrator's ID -> the user names (or ID) of administrators?

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,104 @@
+    }
+
+Next, configure ``conf/shiro.ini`` based on your authentication system. A local database system of `Apache Shiro`_
+is used here to show you a simple example::
```
How about mentioning that a user can also configure Shiro to enable authentication using LDAP or RDBMS as well?

hyangtack replied:
Added

trustin commented on server/src/main/resources/webapp/scripts/components/util/notification-util.service.js:
```diff
@@ -31,11 +31,14 @@ angular.module('CentralDogmaAdmin')
                          });
+                       } else if (arg.status === 403) {
+                         // TODO(hyangtack) Refine the error message.
+                         Notification.error('You have no permission');
```
missing period? or `Permission denied`

trustin commented on server/src/main/resources/webapp/scripts/components/util/confirmation-dialog.js:
```diff
@@ -0,0 +1,68 @@
+      return defer.promise;
+    }
+  }
+});
```
Should end with a newline

hyangtack replied:
Fixed

trustin commented on server/src/main/resources/webapp/scripts/components/util/confirmation-dialog.html:
```diff
@@ -0,0 +1,11 @@
+  <button class="btn btn-primary"
+          ng-click="modalOptions.ok();">{{ 'components.button_yes' | translate }}
+  </button>
+</div>
```
Should end with a newline

hyangtack replied:
Fixed

hyangtack commented:
@trustin I think I addressed every comment you mentioned. Please let me know if there is any comment I missed.  Also, I fixed the followings:

- Fixed UI for project list. Added a list of removed projects and the button which restores each project.
- Allow everyone to create project from now on. Also fixed the document related to the changes. (As you know, only the administrators were able to do it before)

minwoox commented on server/src/main/java/com/linecorp/centraldogma/server/internal/api/MetadataApiService.java:
```diff
@@ -0,0 +1,393 @@
+     *
+     * <p>Updates the {@link PerRolePermissions} of the specified {@code projectName} and {@code repoName}.
+     */
+    @Put("/metadata/{projectName}/repos/{repoName}/perm/role")
```
Put method without the identifier at the end looks weird to me...

How about `@Put("/metadata/{projectName}/rolePermission/{repoName}")`?
I think this is updating the `rolePermission` of the specific repository, so `repoName` should be at the last.
Also I think we do not have to define the path as same with the API updating `perUserPermissions` and `perTokenPermissions` since this has no identifier(Updating the whole permission).
WDYT?

hyangtack replied:
I agree with you. I also think POST is more reasonable. Thanks!

minwoox commented on server/src/main/java/com/linecorp/centraldogma/server/internal/api/ProjectServiceV1.java:
```diff
@@ -93,134 +94,62 @@ public ProjectServiceV1(ProjectManager projectManager, CommandExecutor executor,
-    public CompletionStage<ProjectInfo> removeToken(@Param("projectName") String projectName,
-                                                    @Param("appId") String appId,
+    @Patch("/projects/{projectName}")
+    @Decorator(AdministratorsOnly.class)
```
If the owner of a project removes the project by mistake, does he/she have to ask the administrator to bring it back?

trustin replied:
That's a good question. I wonder if the project owner is allowed to unremove his or her project.

hyangtack replied:
As we discussed in LINE chat room, it'd be better to leave this as it is. A metadata is stored in a `meta` repository of each project so the metadata is not able to be accessed if a project is removed, by default. That means we have no chance to find a role of a user in our authentication decorator, for now. If we want to read a metadata from a removed project, we need to make some code more, but it's not an important specification now.

minwoox commented on server/src/main/java/com/linecorp/centraldogma/server/internal/api/TokenService.java:
```diff
@@ -0,0 +1,156 @@
+                                              @RequestObject Author author,
+                                              @RequestObject User loginUser) {
+        return getTokenOrRespondForbidden(appId, loginUser).thenCompose(
+                token -> mds.destroyToken(author, appId)
```
just wondering is there any reason that you used `destroy` instead of `delete`?

hyangtack replied:
`removeToken()` removes a token from a project. And `destroyToken()` removes a token from the system completely (removing it from every project and the system). So I want to make its meaning clear by using `destroy` instead of `delete`.

trustin commented on common/src/main/java/com/linecorp/centraldogma/internal/Util.java:
```diff
@@ -179,6 +179,19 @@ public static String validateEmailAddress(String emailAddr, String paramName) {
+        if (isValidEmailAddress(emailAddr)) {
+            return emailAddr;
+        }
+        return emailAddr + "@localhost.localdomain";
```
Need to raise an exception if the result is not a valid email address yet.

hyangtack replied:
I think I decided bad name for this method, so I changed it to `toEmailAddress` which means converting an argument `emailAddr` to an email address if it's not.

trustin commented on common/src/main/java/com/linecorp/centraldogma/internal/Util.java:
```diff
@@ -179,6 +179,19 @@ public static String validateEmailAddress(String emailAddr, String paramName) {
+        return emailAddr + "@localhost.localdomain";
+    }
+
+    public static String parseLogin(String emailAddr, String paramName) {
```
How about `emailToUsername`? It doesn't seem to parse but just extract in my opinion.

trustin commented on common/src/main/java/com/linecorp/centraldogma/internal/jsonpatch/PathValueOperation.java:
```diff
@@ -66,6 +66,10 @@
         this.value = value.deepCopy();
     }
 
+    public JsonNode value() {
```
This seems unused?

hyangtack replied:
It's used in the `MetadataApiService` like the following:

```
final ProjectRole role = ProjectRole.of(operation.value());
```

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/internal/api/ProjectServiceV1.java:
```diff
@@ -93,134 +94,62 @@ public ProjectServiceV1(ProjectManager projectManager, CommandExecutor executor,
-
     /**
-     * POST /projects/{projectName}/members/{user}
+     * GET /projects/{projectName}/accessibility
```
The term 'accessibility' is often used for disabled people. Do we have a better term here?

hyangtack replied:
I will remove this API, then add `checkPermissionOnly=true` parameter to `GET /projects/{projectName}` API.

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/internal/api/HttpApiExceptionHandler.java:
```diff
@@ -75,6 +77,10 @@ public HttpResponse handleException(RequestContext ctx, HttpRequest req, Throwab
         }
 
+        if (cause instanceof IllegalAccessException) {
+            return HttpResponse.of(HttpStatus.FORBIDDEN);
```
This seems unnecessary now that we throw an `HttpStatusException`?

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/internal/metadata/Repo.java:
```diff
@@ -0,0 +1,145 @@
+ */
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonInclude(Include.NON_NULL)
+public class Repo implements Identifiable {
```
How about `RepositoryMetadata`?

hyangtack replied:
Ok, it seems more consistent.

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/internal/metadata/RepositoryUtil.java:
```diff
@@ -0,0 +1,180 @@
+        return future;
+    }
+
+    private void push0(String projectName, String repoName, Author author, String commitSummary,
```
Probably OK to name as `push`?

trustin commented on server/src/main/java/com/linecorp/centraldogma/server/internal/metadata/package-info.java:
```diff
@@ -0,0 +1,20 @@
+ */
+
+/**
+ * Metadata management for projects in the Central Dogma.
```
for Central Dogma projects?

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+Authentication and Access Control
+=================================
+
+*Central Dogma* provides the features for authentication and access control.
```
No need to italicize

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+
+.. note::
+
+    `Apache Shiro`_ supports RDBMS or LDAP based security system as well. You can find the example configuration
```
system -> systems

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+Access Control
+--------------
+
+One of the way of configuring the access control system for *Central Dogma* is setting a project with a web UI.
```
- One way to configure
- setting a project with a web UI -> to use the web UI?

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+--------------
+
+One of the way of configuring the access control system for *Central Dogma* is setting a project with a web UI.
+To open the web UI, you may access ``http://{your-central-dogma-domain-or-ip}:36462`` by your web browser.
```
by your -> in your

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+
+One of the way of configuring the access control system for *Central Dogma* is setting a project with a web UI.
+To open the web UI, you may access ``http://{your-central-dogma-domain-or-ip}:36462`` by your web browser.
+You may configure a project with HTTP APIs, but we recommend the web UI because it is easier and simple.
```
easier and simpler

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+You may configure a project with HTTP APIs, but we recommend the web UI because it is easier and simple.
+
+Everyone who is logged in is able to create a new project, and he or she would be an owner of the project.
+If you have a right to configure a project, in other words, if you are an owner of the project,
```
a right -> the right

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+.. image:: _images/auth_1.png
+
+You can see the configuration UI for a project, in the below example it is for the project ``sample``.
+With this UI, you can add a user or token as a member of the project. You can also remove the member or token from the project.
```
- With this UI -> In this page
- a user or a token
- the member of token -> them

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+
+.. image:: _images/auth_2.png
+
+When you add a user as a member of the project, you need to choose a role of the member.
```
a role -> the role?

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+.. image:: _images/auth_2.png
+
+When you add a user as a member of the project, you need to choose a role of the member.
+There are 4 types of user role in the access control system of *Central Dogma*, however you can choose one
```
- 4 user role types
- however -> but

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+
+When you add a user as a member of the project, you need to choose a role of the member.
+There are 4 types of user role in the access control system of *Central Dogma*, however you can choose one
+of the ``Owner`` and ``Member`` role in the UI. More information about the role is as follows.
```
the owner -> owner

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+
+- ``Administrator``
+
+  - the user that every permissions are assigned to. Only the administrator can restore removed project.
```
- every -> all
- are assigned to -> are assigned to, a.k.a. 'super user'.
- the administrator -> an administrator

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+- ``Administrator``
+
+  - the user that every permissions are assigned to. Only the administrator can restore removed project.
+    The administrators can be configured by ``conf/dogma.json`` as described the above.
```
by -> in

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+
+- ``Owner`` of a project
+
+  - the administrator of a project. The owners can create a new repository and remove a repository from the project.
```
- The owners -> Owners ?

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+
+- ``Member`` of a project
+
+  - the user who can see the configuration of a project, but the members is not allowed to change the configuration.
```
the user -> users
is -> are

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+
+- ``Guest`` of a project
+
+  - the user who is logged in but he or she is neither an owner nor a member of a project. The user is not one of
```
- he or she is -> is
- The user is .. -> A guest is not an administrator, ans thus is not allowed to ...

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+.. note::
+
+    Do not forget to make a new ``Application Token`` before adding a token to a project. ``Add a token`` button
+    would be disabled if there is no token.
```
Should we show some message like 'Click here to create a new application token' when no token is available?

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+The following image shows the configuration of the ``main`` repository.
+With this UI, you can do the followings.
+
+- Changing the role of a member or a token in the project
```
the project -> a project

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+
+- Changing the role of a member or a token in the project
+- Setting permissions of each role for a repository
+- Setting permissions of the specific member or token for a repository
```
the specific -> a specific

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+.. image:: _images/auth_3.png
+
+Permissions can be specified for a repository only. So a user can configure their repositories with
+the different access control levels. There are only two types of the permission now, which are ``READ`` and ``WRITE``.
```
- the different -> different
- two permission types currently

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+
+Permissions can be specified for a repository only. So a user can configure their repositories with
+the different access control levels. There are only two types of the permission now, which are ``READ`` and ``WRITE``.
+``WRITE`` permission automatically includes ``READ`` permission, so you cannot give ``WRITE`` permission only
```
- automatically includes -> implies
- give WRITE permission only -> give only WRITE permission (for disambiguation, just in case a user interprets as 'only to a user')

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+Permissions can be specified for a repository only. So a user can configure their repositories with
+the different access control levels. There are only two types of the permission now, which are ``READ`` and ``WRITE``.
+``WRITE`` permission automatically includes ``READ`` permission, so you cannot give ``WRITE`` permission only
+to an user, a token or any role.
```
an user -> a user

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+to an user, a token or any role.
+
+Every access of HTTP API will be controlled by the access control system. A request is allowed only if the
+user of the request has the right to do the action. If permissions for the user are specified in the repository
```
the right to do the action -> sufficient permissions.

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+Application Token
+^^^^^^^^^^^^^^^^^
+
+When a user uses the web UI, he or she should login first to get a token for the web session. But what should a user
```
the web session -> a new web session

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+^^^^^^^^^^^^^^^^^
+
+When a user uses the web UI, he or she should login first to get a token for the web session. But what should a user
+do when he or she uses one of *Central Dogma* clients? The user may login *Central Dogma* server through HTTP login API
```
login... -> log in to a Central Dogma server via HTTP API

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+
+``Application Token`` is like a virtual user, so it can have any role in a project. Also, its permissions can be specified
+in a repository configuration like a member. To get a new token, a user can use ``Application Tokens`` menu of the web UI.
+``Application ID`` is should be unique to identify where the request is from.
```
- is should be -> has to be
- the request is from -> a client request comes from

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+
+.. image:: _images/auth_4.png
+
+Anyone who is logged in the *Central Dogma* can create a new ``Application Token``, and the token is shared
```
logged into

trustin commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,133 @@
+
+Anyone who is logged in the *Central Dogma* can create a new ``Application Token``, and the token is shared
+for everyone. So any owner of a project can add any token to their project. However only both the token creator
+and the administrator are allowed to inactivate and/or remove the token.
```
inactivate -> deactivate

hyangtack commented on server/src/main/java/com/linecorp/centraldogma/server/internal/admin/authentication/UserWithToken.java:
```diff
@@ -18,18 +18,32 @@
+import com.linecorp.centraldogma.server.internal.metadata.Token;
+
+/**
+ * A {@link User} which accesses the API with a secret.
```
`secret` is correct here because API requests will be sent with a `secret` and a token will be found by the `secret`.

minwoox commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,138 @@
+
+When a user uses the web UI, he or she should login first to get a token for a new web session. But what
+should a user do when he or she uses one of Central Dogma clients? The user may log in to a Central Dogma
+server via HTTP API and gets a session token. But it is inconvenient and the user may write more complicated
```
gets -> get? (because `may` is auxiliary verb here)

hyangtack commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,138 @@
+.. note::
+
+    Do not forget to make a new ``Application Token`` before adding a token to a project. ``Add a token``
+    button would be disabled if there is no token. The pencil icon on the right of the ``Tokens`` title
```
It's a title in a web UI.

hyangtack commented on site/src/sphinx/auth.rst:
```diff
@@ -0,0 +1,138 @@
+
+Permissions can be specified for a repository only. So a user can configure their repositories with different
+access control levels. There are only two permission types currently, which are ``READ`` and ``WRITE``.
+``WRITE`` permission implies ``READ`` permission, so you cannot give only WRITE permission to a user,
```
Left this as it is because we discussed about this last week.

hyangtack commented:
Thanks, reviewers!
